runtime/common/src/fees.rs

```diff
@@ -0,0 +1,460 @@
+            // assume fee_factor is 20_000_000_000, then you would get (20_000_000_000 / 10_000_000_000 =) 2 units per ZTG
+            let converted_fee = bmul(native_fee, fee_factor)
+                .ok_or(TransactionValidityError::Invalid(InvalidTransaction::Custom(0u8)))?;
+
+            Ok(converted_fee)
+        }
+
+        #[cfg(feature = "parachain")]
+        pub(crate) fn get_fee_factor(
+            asset_id: CurrencyId,
+        ) -> Result<Balance, TransactionValidityError> {
+            let location = AssetConvert::convert(asset_id);
+            let metadata = location
+                .and_then(|loc| {
+                    <AssetRegistry as orml_traits::asset_registry::Inspect>::metadata_by_location(
+                        &loc,
+                    )
+                })
+                .ok_or(TransactionValidityError::Invalid(InvalidTransaction::Custom(2u8)))?;
+            let fee_factor = metadata
+                .additional
+                .xcm
+                .fee_factor
+                .ok_or(TransactionValidityError::Invalid(InvalidTransaction::Custom(3u8)))?;
+            Ok(fee_factor)
+        }
+
+        pub struct TTCBalanceToAssetBalance;
+        impl BalanceConversion<Balance, CurrencyId, Balance> for TTCBalanceToAssetBalance {
+            type Error = TransactionValidityError;
+
+            fn to_asset_balance(
+                native_fee: Balance,
+                asset_id: CurrencyId,
+            ) -> Result<Balance, Self::Error> {
+                match asset_id {
+                    Asset::Ztg => Ok(native_fee),
+                    #[cfg(not(feature = "parachain"))]
+                    Asset::ForeignAsset(_) => {
+                        return Err(TransactionValidityError::Invalid(InvalidTransaction::Custom(
+                            1u8,
+                        )));
+                    }
+                    #[cfg(feature = "parachain")]
+                    Asset::ForeignAsset(_) => {
+                        let fee_factor = get_fee_factor(asset_id)?;
+                        let converted_fee = calculate_fee(native_fee, fee_factor)?;
+                        Ok(converted_fee)
+                    }
+                    Asset::CategoricalOutcome(_, _)
+                    | Asset::ScalarOutcome(_, _)
+                    | Asset::CombinatorialOutcome
+                    | Asset::PoolShare(_) => {
+                        return Err(TransactionValidityError::Invalid(InvalidTransaction::Custom(
+                            2u8,
+                        )));
+                    }
+                }
+            }
+        }
+
+        pub struct TTCHandleCredit;
+        impl HandleCredit<AccountId, Tokens> for TTCHandleCredit {
+            fn handle_credit(final_fee: CreditOf<AccountId, Tokens>) {
+                // Handle the final fee and tip, e.g. by transferring to the treasury.
+                DealWithForeignFees::on_unbalanced(final_fee);
```

Comment:
AFAIK if the `asset_id` is not `None`, this function is invoked.

Two issues I see with that:
1. If the `asset_id` is `Asset::Ztg`, it seems to use `DealWithForeignFees` instead of `DealWithFees`, which would be called in case the `asset_id` is `None` which should be equal to `Asset::Ztg`. Plus both of those implementations could differ.
2. The `orml-tokens` pallet cannot handle `Asset::Ztg`, as it is not associated to the native token. You can try that out by querying the ZTG amount of one account using a storage query into `system.account` and then trying to query the same account using `tokens.account` using the same account and the asset id `Ztg`. The `orml-currencies` pallet acts as a intermediary here. Should it receive `Asset::Ztg` as the asset id, it will automatically route the request to `pallet-balances`. In any other case, it will route the request to `pallet-tokens`.

Reply:
> I thought if nobody uses it anyways... I doesn't really matter here.

I don't think we should make this implication, as we don't know the intentions of everybody who is ever going to interact with the protocol in that way.
Anybody who looks up the `Asset` type that is used as an argument here will see that `ZTG` is a variant and might be tempted to use it - either unknowingly, on purpose for the sake of ambiguousness or for any other reason.

I think the cleanest way would be to route it to the same path that native ZTG is using, as in fact, it is used in prediction markets and swaps to represent the real ZTG and mapped to the native currency via `orml-currencies`. If there are strong reason not to do that, I would be open to discuss throwing an error that refers to using `None` instead.

Reply:
pallet_transaction_payment abstracts this away. Routing to from the pallet_asset_tx_payment to the pallet_transaction_payment is not possible without having a hard fork. If the user specified `None` it uses the real ZTG of pallet_balances. I see, we should probably better throw an error for this case. (as you outlined)

Reply:
c5ee57bc0eeb2286937a5ef7607d279f57448b35